taming-the-stack: Add session content (#43)
open-education-hub/essentials-security: taming-the-stack/index.md

## Stack

### An abstract data structure

The stack, as a data structure, functions on the LIFO (Last In First Out) concept: one can access only one element at a time, and it's always the last one.

The two operations accepted by the stack data structure are:

- `push` (add one element at the back)
- `pop` (retrieve the last element, while removing it from the stack)

### Real life usecase

Enough about abstract data structures for now, let's get down to the OS business!
Every process there is on your system has a stack, used for a plethora of reasons.
While the program stack has a similar behaviour with the abstract data structure from which it got its name, its role is of great interest to attackers and programmers alike:

- **it stores the return address**

   Have you ever wondered how does a program know how to return and execute the next instruction after a function call?!
   Here comes the stack.
   The return address (meaning the address of the next instruction after a function call) gets **pushed** on the stack right before entering the function and **popped** at the end of the function.

   ```asm
   call my_func
   
   ; this is equivalent to

   inc rip
   push rip

>    jmp my_func

Reviewer: Explaining this via text is a bit bland. You can use a GIF alongside some shorter text. You can make GIFs from SVG files using [draw.io](https://app.diagrams.net/) and `ffmpeg`. We are using this for the OS class [1]. Albeit a bit lacklustre, this method is the best we've come up with so far.

[1] https://github.com/open-education-hub/operating-systems/blob/master/content/common/makefile/slides.mk#L21-L25

   For instance, the following C function gets translated to Assembly like so:

   ```C
   void my_func()
   {
        int a = 5;
   }
   ```

   ```asm
   my_func:
        ; preserve the stack frame, a special register
        push ebp
        mov ebp, esp

        ; allocate space on stack for an integer
        sub esp, 4
        ; initialize the integer with 5
        mov [esp], 5

        ; restore the stack pointer and the stack frame
        mov esp, ebp
        pop ebp
        ret
   ```

### `push` and `pop` under the hood

As you probably already figured, the fact that the stack operates with its last added value only means that we somehow need to store the exact address of the top somewhere.
Here comes the `esp` register.
Whenever a `push` or `pop` occurs, this register either gets decreased or increased with the size of the value stored on the stack.
You've read it righ: the **`push` operation decreases the `esp` value** and a **`pop` operation increases it**!
That's because the stack grows downwards.

![stack growth](assets/stack-high-low.png)

So, the following 4 snippets are equivalent 2 by 2:

```asm
push dword 5
```

```asm
sub esp, 4

> mov [esp], 5

Reviewer: ```suggestion
mov [esp], dword 5
```

```

```asm
pop ecx
```

```asm
mov [esp], ecx
add esp, 4
```

> ## Functions and the stack

Reviewer: This entire section requires a GDB demo.

Every function has two classes of values, usually stored on stack, extremely important for its well-being:

1. the return address

1. the parameters / arguments

Meddling with these might get you to a big fat **SEGFAULT** or to great power.

### `ebp`, the stack frame

But before discussing that, we have to bring light to another obscure register, `ebp`.
We kind of used it before, in our journey, as it has a great advantage.
It stores the stack pointer value right before the stack begins to hold local variables and preserved register values.
In other words, it keeps a pointer to the stack at the beginning of the function, enabling us to actually move freely through the stack.
We will, now, refer to values stored on it, even though they are not the last ones.

```asm
push ebp
mov ebp, esp

push dword 3
push dword 4
push dword 5

; at this point esp decreased its value with 3 * 4 = 12 bytes
; traditionally we can access the last value only,
; however the stack is like an array, so we will use the pointers
; it offers us

mov eax, [esp + 8] ; eax = 3
mov eax, [ebp - 4] ; eax = 3
```

### The return address

> The return address of a function is THE MOST TARGETED information in an attack.

Reviewer: ```suggestion
The return address of a function is **the most targeted** information in an attack.
```
Bold is more visually appealing than CAPS LOCKING.

And secondly, I would tune this statement down. Say it's one of the most targeted pieces of information. There are plenty of other things one can attack and IDK what the SotA is regarding binary exploits.

> Literally.

Reviewer: Ditto: chill.

> There is even a special class of attacks that takes its name from it, ROP (Return Oriented Programming).

Reviewer: Link the ROP session of the binary track: https://security-summer-school.github.io/binary/return-oriented-programming/

> The reason for this kind of popularity is obvious: it represents one of the rare instances when the program **performs a jump to an address saved on stack**, which, combined with the stupidiy or the laziness of the programmer, can result in a nasty backdoor to the system.

Reviewer: I think you can just define and use the concept of _code pointer_. It's just more straightforward like so.

The address at which the return address is usually stored on x86 systems is `[ebp + 4]`.

### The parameters

> The parameters follow a similar story to the one of the return address, with a slight modification, though.

Reviewer: ```suggestion
The parameters follow a similar story to that of the return address, with a slight modification, though.
```

On 64-bit x86 they are placed in special registers, if possible.
If the number of parameters is high, they would get transmitted using the stack, just as it happens, on 32-bit x86.

The address at which the first parameter gets stored on 32-bit x86 systems is `ebp + 8`.

The address at which the second parameter gets stored on 32-bit x86 systems is `ebp + 12`.

The address at which the third parameter gets stored on 32-bit x86 systems is `ebp + 16`.

And so on.

![parameters and ebp](assets/function-stack.jpg)

## Summary

> The both curse and blessing of modern C/C++ code is the absolute control it gives the programmer.

Reviewer: ```suggestion
Modern C/C++ code is both a blessing and a curse because it gives the programmer absolute control over the memory.
```
